tools/aot/aot_compile.bzl

```diff
@@ -3,29 +3,243 @@
+                requirement("numpy"),
+                requirement("torch"),
+                requirement("torch-mlir"),
+                "//tools/aot:torch_loader_utils",
+            ],
+            # This is needed for testing the binary standalone
+            args = ["--torch_loader_path=" + torch_loader_path],
+        )
+
+        native.genrule(
+            name = "gen_" + name + "_reference_tensors",
+            srcs = [],
+            outs = [reference_tensors_file],
+            cmd = "./$(location " + torch_exporter + ")" +
+                  " --torch_loader_path=" + torch_loader_path +
+                  " --reference_tensors_path=$(location " + reference_tensors_file + ")",
+            tools = [torch_exporter],
+        )
+
+        native.genrule(
+            name = "gen_" + name + "_mlir_torch",
+            srcs = [],
+            outs = [_name + "_torch.mlir"],
+            cmd = "./$(location " + torch_exporter + ")" +
+                  " --torch_loader_path=" + torch_loader_path +
+                  " > $(OUTS)",
+            tools = [torch_exporter],
+        )
+
+        native.genrule(
+            name = "gen_" + name + "_mlir_tcp",
+            srcs = [_name + "_torch.mlir"],
+            outs = [_name + "_tcp.mlir"],
+            cmd = "./$(location //:tcp-opt)" +
+                  " -torch-backend-to-tcp-backend-pipeline $(SRCS)" +
+                  " > $(OUTS)",
+            tools = ["//:tcp-opt"],
+        )
+
+    native.genrule(
+        name = "gen_" + name + "_mlir_llvm",
+        # When tcp_source is provided, prefer that as the start for aot_compile;
+        # else continue using genrule generated *_tcp.mlir (torch_export workflow)
+        srcs = [tcp_source or (_name + "_tcp.mlir")],
+        outs = [_name + "_llvm.mlir"],
+        cmd = "./$(location //:tcp-opt)" +
+              " -tcp-to-llvm-pipeline $(SRCS)" +
+              " > $(OUTS)",
+        tools = ["//:tcp-opt"],
+    )
+
     native.genrule(
-        name = "_internal_gen_asm_" + name,
-        srcs = [tcp_source],
-        outs = ["_internal_" + name + ".S"],
-        cmd = "./$(location //:tcp-opt) -tcp-to-llvm-pipeline $(SRCS) | ./$(location @llvm-project//mlir:mlir-translate) -mlir-to-llvmir | ./$(location @llvm-project//llvm:llc) -O3 > \"$@\"",
-        tools = [
-            "//:tcp-opt",
-            "@llvm-project//mlir:mlir-translate",
-            "@llvm-project//llvm:llc",
-        ],
+        name = "gen_" + name + "_llvm_ir",
+        srcs = [_name + "_llvm.mlir"],
+        outs = [_name + ".ll"],
+        cmd = "./$(location @llvm-project//mlir:mlir-translate)" +
+              " -mlir-to-llvmir $(SRCS)" +
+              " > $(OUTS)",
+        tools = ["@llvm-project//mlir:mlir-translate"],
     )
 
-    native.cc_library(
+    native.genrule(
+        name = "gen_" + name + "_host_asm",
+        srcs = [_name + ".ll"],
+        outs = [_name + ".S"],
+        cmd = "./$(location @llvm-project//llvm:llc) -O3 < $(SRCS)" +
```

Comment:
![image](https://github.com/cruise-automation/mlir-tcp/assets/19234106/0000cafa-47c1-4c8e-b25f-2c6834f609a3)
https://github.com/cruise-automation/mlir-tcp/commit/ada575353cd2381504447c93bc204773134a9fc2#r139404324

Reply:
<img width="791" alt="image" src="https://github.com/cruise-automation/mlir-tcp/assets/19234106/e65cfbe6-800c-4e99-92d0-eee6d736f16d">
https://github.com/cruise-automation/mlir-tcp/commit/ada575353cd2381504447c93bc204773134a9fc2#r139411511

Reply:
@AaronStGeorge I did see your internal PR to replace llc with clang, and I think it is a good move for AOT compile at a later point when we do want to position this as a "compiler generating optimized machine code". Right now it isn't intended to, and very explicitly so - the `cc_library` target for `aot_compiled_foo` is marked `testonly` - this will error when added as a dependency for a non-test targets. The goal is to avoid a user from inadvertently treating the compiled artifacts as production grade / optimized. I'll update the README to emphasize this, and add a note about replacing llc with clang as a potential future TODO when we're ready to expose this is a true compiler.

Apart from missed optimizations (which is okay for e2e testing), do you see any other concerns with using `llc`?

Reply:
> I'll update the README to emphasize this, and add a note about replacing llc with clang as a potential future TODO when we're ready to expose this is a true compiler.

Done (https://github.com/cruise-automation/mlir-tcp/pull/46/commits/e01766d4c9d7f922dd28a5a8bdf7caf12f581bdb).

Reply:
👍 with more context, choosing `llc` makes a lot of sense.